tracee-rules/engine/engine.go

```diff
@@ -68,8 +68,8 @@ func NewEngine(sigs []types.Signature, sources EventSources, output chan types.F
 func (engine *Engine) Start(done chan bool) {
 	go engine.consumeSources(done)
 	for s, c := range engine.signatures {
-		defer close(c)
 		go func(s types.Signature, c chan types.Event) {
+			defer close(c)
```

Comment:
what's the motivation for this change? the current state feels more reasonable to me

Reply:
The current implementation will panic if an incorrectly written signature erroneously sends an event on the channel after signalling `done`. Here's a test to show this behaviour: https://gist.github.com/simar7/d9ec8ef5df938b1b8a4dc86d692f5076

In addition, `c` is a channel that's read by the engine (Start func acts as a consumer) and created by someone else https://github.com/aquasecurity/tracee/blob/efdc613024f04f45b4be877372bd474f31da890d/tracee-rules/main.go#L34-L39

So one can also debate that we shouldn't close at all, which is something that crossed my mind as I write this. 

What do you think?

Reply:
> incorrectly written signature sends an event

signatures don't send events, they receive them. only input sources send events (in package main). so for your test to repro IRL main needs to have a bug which it closed the engine and then sent an event to the closed engine.

> panic if ... sends an event on the channel after signalling done
this behavior sounds good to me. done means the engine is done and should not process any further events. perhaps an error would be better than panic, but given both main and engine are under our control, and this scenario is so unlikely to repro, I don't see sufficient motivation to improve it.

Also - moving the defer to the goroutine is equivalent to removing it altogether. the defer will never be called where you put it, which is why it doesn't panic in your test.

> c is a channel that's read by the engine (Start func acts as a consumer) and created by someone else

actually this is not the case.. I think you are confusing `c`, which is the signature's dedicated channel and is [created by engine](https://github.com/aquasecurity/tracee/blob/4fdcba8ad7ad51f7bff77faed1add657ecbbf2fa/tracee-rules/engine/engine.go#L34) and owned by it, with `engine.inputs.xxx` channel, which indeed is created by main and isn't owned by engine.

Reply:
> > incorrectly written signature sends an event
> 
> signatures don't send events, they receive them. only input sources send events (in package main). so for your test to repro IRL main needs to have a bug which it closed the engine and then sent an event to the closed engine.
> 

Correct. I do agree the chances of that happening are low as we should limit the amount of logic we write in main, as it's hard to test for.

> > panic if ... sends an event on the channel after signalling done
> > this behavior sounds good to me. done means the engine is done and should not process any further events. perhaps an error would be better than panic, but given both main and engine are under our control, and this scenario is so unlikely to repro, I don't see sufficient motivation to improve it.
> 
> Also - moving the defer to the goroutine is equivalent to removing it altogether. the defer will never be called where you put it, which is why it doesn't panic in your test.
> 

You're right - I didn't realize this.

> > c is a channel that's read by the engine (Start func acts as a consumer) and created by someone else
> 
> actually this is not the case.. I think you are confusing `c`, which is the signature's dedicated channel and is [created by engine](https://github.com/aquasecurity/tracee/blob/4fdcba8ad7ad51f7bff77faed1add657ecbbf2fa/tracee-rules/engine/engine.go#L34) and owned by it, with `engine.inputs.xxx` channel, which indeed is created by main and isn't owned by engine.

Gotcha. I understand it now. I'll revert this as well.